tests/add_test.py

```diff
@@ -0,0 +1,228 @@
+        line = line.strip()
+        if (disasm_info.entry + b" ")[2:].upper() in line:
+            output.append(b"glabel test")
+
+        index = len(output)
+        if line.startswith(b"func"):
+            line = b"glabel " + line[:-1]
+        elif b"lui" in line:
+            reg, imm_hex = line.split(b" ")[-2::]
+            imm = int(imm_hex, 0)
+            if 0x10 <= imm < 0x1000:  # modestly large lui is probably a pointer
+                waiting_hi[reg[:-1]] = (index, imm, imm_hex)
+        else:
+            lo_regs = [reg for reg in waiting_hi.keys() if reg in line]
+            if lo_regs and "0x" in line:
+                lo_reg = lo_regs[0]
+                hi_ind, hi_imm, hi_imm_hex = waiting_hi[lo_reg]
+                lo_imm_hex = line.split(b" ")[-1].split(b"(")[0]
+                lo_imm = int(lo_imm_hex, 0)
+                if lo_imm >= 0x8000:
+                    lo_imm -= 0x10000
+                sym = b"D_" + bytes(hex((hi_imm << 16) + lo_imm)[2:].upper(), "utf-8")
+                line = replace_last(line, lo_imm_hex, b"%lo(" + sym + b")")
+                output[hi_ind] = replace_last(
+                    output[hi_ind], hi_imm_hex, b"%hi(" + sym + b")"
+                )
+                del waiting_hi[lo_reg]
+        output.append(line)
+    return b"\n".join(output)
+
+
+def irix_compile_with_flag(in_file: Path, out_file: Path, flag: str) -> None:
+    logging.info(f"Compiling {in_file} to {out_file} using this flag: {flag}")
+    with ExitStack() as stack:
+        temp_o_file = stack.enter_context(NamedTemporaryFile(suffix=".o")).name
+        temp_out_file = stack.enter_context(NamedTemporaryFile(suffix=".out")).name
+        logging.debug(f"Compiling and linking {in_file} with {flag}...")
+        do_compilation_step(temp_o_file, str(in_file), flag)
+        do_linker_step(temp_out_file, temp_o_file)
+        disasm_info = do_disassembly_step(temp_out_file)
+        final_asm = do_fix_lohi_step(disasm_info)
+    out_file.write_bytes(final_asm)
+    logging.info(f"Successfully wrote disassembly to {out_file}.")
+
+
+def add_test_from_file(orig_file: Path) -> None:
+    test_dir = Path("end_to_end") / orig_file.stem
+    try:
+        test_dir.mkdir()
+    except FileExistsError:
+        raise Exception(f"{test_dir} already exists. Name your test something else.")
+    logging.debug(f"Created new directory: {test_dir}")
+
+    in_file: Path = test_dir / "orig.c"
+    shutil.copy(str(orig_file), str(in_file))
+    logging.debug(f"Created {in_file}")
+
+    for asm_filename, flag in OUT_FILES_TO_IRIX_FLAG.items():
+        asm_file_path = Path(str(test_dir / asm_filename) + ".s")
+        irix_compile_with_flag(in_file, asm_file_path, flag)
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(
+        description="Add or update end-to-end decompiler tests."
+    )
+    parser.add_argument(
+        "files", help="files containing C code to compile (then decompile)", nargs="+"

```

Comment:
What's the expected form of these paths? `./tests/<name>/orig.c`? `any-dir/<name>.c`?

Reply:
Pretty much `any-dir/<name>.c`. Should probably write this expectation explicitly.

Reply:
I think requiring `tests/<name>/orig.c` might be preferable, since:
a) it makes add and update workflows closer
b) it means you don't have to create a temporary file and then remove it to add a new test
c) if there are two copies of the same file floating around (both the original and orig.c), you risk accidentally modifying the wrong file